modules/fedimint-lnv2-common/src/contracts.rs

```diff
@@ -0,0 +1,170 @@
+        expiration: u64,
+        claim_pk: PublicKey,
+        refund_pk: PublicKey,
+        ephemeral_pk: PublicKey,
+    ) -> Self {
+        let commitment = Commitment {
+            payment_hash: preimage.consensus_hash(),
+            amount,
+            expiration,
+            claim_pk,
+            refund_pk,
+            ephemeral_pk,
+        };
+
+        let ciphertext = encrypt_preimage(
+            &agg_pk,
+            &encryption_seed,
+            &preimage,
+            &commitment.consensus_hash(),
+        );
+
+        IncomingContract {
+            commitment,
+            ciphertext,
+        }
+    }
+
+    pub fn contract_id(&self) -> ContractId {
+        ContractId(self.consensus_hash())
+    }
+
+    pub fn verify(&self) -> bool {
+        verify_ciphertext(&self.ciphertext, &self.commitment.consensus_hash())
+    }
+
+    pub fn verify_decryption_share(
+        &self,
+        pk: &PublicKeyShare,
+        dk_share: &DecryptionKeyShare,
+    ) -> bool {
+        verify_decryption_key_share(
+            pk,
+            dk_share,
+            &self.ciphertext,
+            &self.commitment.consensus_hash(),
+        )
+    }
+
+    pub fn verify_agg_decryption_key(
+        &self,
+        agg_pk: &AggregatePublicKey,
+        agg_decryption_key: &AggregateDecryptionKey,
+    ) -> bool {
+        verify_agg_decryption_key(
+            agg_pk,
+            agg_decryption_key,
+            &self.ciphertext,
+            &self.commitment.consensus_hash(),
+        )
+    }
+
+    pub fn verify_preimage(&self, preimage: &[u8; 32]) -> bool {
+        preimage.consensus_hash::<sha256::Hash>() == self.commitment.payment_hash
+    }
+
+    pub fn decrypt_preimage(
+        &self,
+        agg_decryption_key: &AggregateDecryptionKey,
+    ) -> Option<[u8; 32]> {
+        let preimage = decrypt_preimage(&self.ciphertext, agg_decryption_key);
+
+        match self.verify_preimage(&preimage) {
+            true => Some(preimage),
+            false => None,
+        }
+    }
+
+    pub fn create_decryption_key_share(&self, sk: &SecretKeyShare) -> DecryptionKeyShare {
+        create_decryption_key_share(sk, &self.ciphertext)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, Encodable, Decodable)]
+pub struct OutgoingContract {
+    pub payment_hash: sha256::Hash,
+    pub amount: Amount,
+    pub expiration: u64,
+    pub claim_pk: PublicKey,
+    pub refund_pk: PublicKey,
+    pub ephemeral_pk: PublicKey,
+    pub invoice_hash: sha256::Hash,
+}
+
+impl OutgoingContract {
+    pub fn contract_id(&self) -> ContractId {
+        ContractId(self.consensus_hash())
+    }
+
+    pub fn forfeit_message(&self) -> Message {
+        Message::from(self.contract_id().0)
+    }
+
+    pub fn verify_preimage(&self, preimage: &[u8; 32]) -> bool {
+        preimage.consensus_hash::<sha256::Hash>() == self.payment_hash
+    }
+
+    pub fn verify_forfeit_signature(&self, signature: &Signature) -> bool {
```

Comment:
nit: you could return `Result` here and just to ? when calling instead of checking a boolean